packages/nitro-protocol/contracts/AssetHolder.sol

```diff
@@ -139,203 +139,120 @@ contract AssetHolder is IAssetHolder {
     // Internal methods
     // **************
 
+    function _computeNewAllocation(
+        uint256 initialHoldings,
+        Outcome.AllocationItem[] memory allocation,
+        uint256[] memory indices
+    )
+        internal
+        pure
+        returns (
+            Outcome.AllocationItem[] memory newAllocation,
+            bool safeToDelete,
+            uint256[] memory payouts,
+            uint256 totalPayouts
+        )
+    {
+        payouts = new uint256[](indices.length > 0 ? indices.length : allocation.length); // [] means "all"; values default to 0
+        totalPayouts = 0;
+        newAllocation = new Outcome.AllocationItem[](allocation.length);
+        safeToDelete = true; // switched to false if there is an item with amount > 0
+        uint256 surplus = initialHoldings; // virtual funds available during calculation
+        uint256 k = 0; // indexes indices
+
+        // loop over allocations and decrease surplus
+        for (uint256 i = 0; i < allocation.length; i++) {
+            // copy destination part
+            newAllocation[i].destination = allocation[i].destination;
+            // compute new amount part
+            uint256 affordsForDestination = (allocation[i].amount > surplus)
+                ? surplus
+                : allocation[i].amount; // min(amount,surplus)
+            if ((indices.length == 0) || ((k < indices.length) && (indices[k] == i))) {
```

Comment:
required: The correctness here seems to rely on `indices` being sorted.

Can you confirm that? If so, I think we should assert that `indices` is sorted, when present.

Reply:
The behaviour of the function certainly depends on the ordering of `indices` in this implementation. There is a test case that shows this:
https://github.com/statechannels/statechannels/blob/3391f1fd482e511e16f38b9a79da71faadc14a1c/packages/nitro-protocol/test/contracts/AssetHolder/transfer.test.ts#L59-L61

I think we have two options: 

a) your suggestion: revert if unsorted `indices` is provided. This will cost some gas to do the check,
b) warn in the documentation about providing unsorted `indices`: 

> if at some point `indices[k+1] <= indices[k]`, then all indices >= k + 1 will be effectively ignored. 

This will not cost any extra gas.

Reply:
I implemented a) in 20153dd and 4e243df so we can make a judgement call based on the extra gas spent and either keep or revert/omit those commits.

Reply:
Before: from [here](https://app.circleci.com/pipelines/github/statechannels/statechannels/10879/workflows/f9b7c266-56bc-41a4-93da-6717b3258b5e/jobs/53260/artifacts)
```
13. -> 2 indices:
59975 gas

14. -> 3 indices:
48208 gas
```

After https://github.com/statechannels/statechannels/commit/4e243df18c5ef0b4976647a03731a7a9390a177c, from [here](https://app.circleci.com/pipelines/github/statechannels/statechannels/10936/workflows/3f3bdd67-0bc3-4305-9ee2-2acbe5716190/jobs/53492/artifacts)
```
13. -> 2 indices:
59762 gas

14. -> 3 indices:
47869 gas

15. -> reverse order (see 13):
66606 gas
```

It seems as if the gas consumption _went down_. I assume from this that there is some variance, and that the additional require adds less gas than the variance between test runs?

Since the sorting order of the indices appears vital from a protocol security point of view (right? see below) I think we need to accept whatever the gas cost is, which is fortunately small.

### Percieved problem

If the `indices` array that is passed is not sorted, does it allow one to extract funds that one doesn't deserve? IE. is the order of the outcome items still respected? I could not immediately tell. I had the sense that the order is still preserved, but that the end effect was unintended (only a subset was withdrawn corresponding to a particular subset of the indices that are sorted).

Reply:
Firstly about the gas usage. There should not be any variance, in my understanding, although I do see some small (10 or 15 gas say) variance on circle. I don't currently know what to attribute this to; it might be linked somehow to running chain tests concurrently. 

The first commit you picked is ahead of an unrelated change which I would expect to _also_ reduce gas. 

I think the gas usage went up (as it has to), but it's not easy to glean that from my commit history because the tests didn't always run. I reran this test file on its own locally and got this (increases shown inline)

```
without:
  1. funded          -> 1 EOA:
20977 gas 
 2. overfunded      -> 1 EOA:
26966 gas
 3. underfunded     -> 1 EOA:
28055 gas
 4. funded      -> 1 channel:
32947 gas
 5. overfunded  -> 1 channel:
47947 gas
 6. underfunded -> 1 channel:
49036 gas
 7. -> 2 EOA         1 index:
45066 gas
 8. -> 2 EOA         1 index:
30050 gas
 9. -> 2 EOA         partial:
45130 gas
10. -> 2 chan             no:
34801 gas
11. -> 2 chan           full:
51175 gas
12. -> 2 chan        partial:
66255 gas
13. -> 2 indices:
59750 gas
14. -> 3 indices:
47804 gas
with:
 1. funded          -> 1 EOA:
21018 gas (+41)
 2. overfunded      -> 1 EOA:
27024 gas (+58)
 3. underfunded     -> 1 EOA:
28125 gas (+70)
 4. funded      -> 1 channel:
33017 gas (+70)
 5. overfunded  -> 1 channel:
48017 gas (+70)
 6. underfunded -> 1 channel:
49106 gas (+70)
 7. -> 2 EOA         1 index:
45136 gas (+70)
 8. -> 2 EOA         1 index:
30120 gas (+70)
 9. -> 2 EOA         partial:
45200 gas (+70)
10. -> 2 chan             no:
34871 gas (+70)
11. -> 2 chan           full:
51245 gas (+70)
12. -> 2 chan        partial:
66325 gas (+70)
13. -> 2 indices:
59999 gas (+249)
14. -> 3 indices:
48232 gas (+428)
```

Either way I think the gas increase is quite small and is tolerable.

Reply:
About your perceived problem: I stand by what I put as the [suggested documentation](https://github.com/statechannels/statechannels/pull/3011#discussion_r538174606).

The algorithm walks along the allocation and "virtually" pays out to each destination, in order. If it hits the first supplied index, it will "actually" pay it out. It will then continue walking, without resetting to the start of the allocation, and look for the second index. Once it reaches the end of the allocation, the algorithm terminates. If the second supplied index is not strictly greater than the first, the algorithm will never find it and no additional "actual" payouts will happen (and similarly no additional changes to the allocation will be made). If the second index _is_ greater than the first, the algorithm will find it, but we can make the same argument for the third index (and so on). 

So my claim is that current implementation will payout `indices[i]` for `i=0,1,...,m` where `m` is the largest integer such that `indices[m] > indices[m-1] > ... indices[0]`.

Reply:
I'm comfortable with the suggested documentation!

Reply:
> About your perceived problem: I stand by what I put as the [suggested documentation](https://github.com/statechannels/statechannels/pull/3011#discussion_r538174606).
> 
> The algorithm walks along the allocation and "virtually" pays out to each destination, in order. If it hits the first supplied index, it will "actually" pay it out. It will then continue walking, without resetting to the start of the allocation, and look for the second index. Once it reaches the end of the allocation, the algorithm terminates. If the second supplied index is not strictly greater than the first, the algorithm will never find it and no additional "actual" payouts will happen (and similarly no additional changes to the allocation will be made). If the second index _is_ greater than the first, the algorithm will find it, but we can make the same argument for the third index (and so on).
> 
> So my claim is that current implementation will payout `indices[i]` for `i=0,1,...,m` where `m` is the largest integer such that `indices[m] > indices[m-1] > ... indices[0]`.

Perhaps the algorithm deserves some high-level description like this?